sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,97 @@ def _sage_(self):
                                     hold=True)
         return f
 
+    def principal_value(self):
+        """
+        Compute the Cauchy Principal Value of the definite integral of a real function in the given interval
+        on the real axis.
+        In mathematics, the Cauchy principal value, is a method for assigning values to certain improper
+        integrals which would otherwise be undefined.
+
+        Examples
+        ========
+
+        >>> from sympy import Dummy, symbols, integrate, limit, oo
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x')
+
+        >>> g = x + 1
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        nan
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        nan
+
+        This happens so because, the method principal_value returns -oo for Integral from -oo to 0
+        and +oo for Integral from 0 to oo. While adding both the Integrals i.e. -oo + oo we do not
+        get 0 (as we expect) but nan since sympy realises that -oo and +oo aren't actual numbers that
+        can be added/subtracted.
+
+        >>> Integral(f, (x, -oo, 0)).principal_value()
+        -oo
+
+        >>> Integral(f, (x, 0, oo)).principal_value()
+        oo
+
+        >>> Integral(f, (x, 1, 2)).principal_value()
+        3/8
+
+        >>> Integral(f, (x, -2, -1)).principal_value() + Integral(f, (x, 1, 2)).principal_value()
+        0
+
+        >>> Integral(f, (x, 0.4, 3)).principal_value()
+        3.06944444444444
+
+        >>> Integral(f, (x, -3, -0.4)).principal_value() + Integral(f, (x, 0.4, 3)).principal_value()
+        0
+
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        x, a, b = self.limits[0]
+
+        r = Dummy('r')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a < s < b]
+
+        f = self.function
+        F = integrate(f, x)
+
+        if len(singularities_list) == 0:
+            I = limit(F, x, b, '-') - limit(F, x, a, '+')
```

Comment:
This is not correct if the limits are `-oo` and `oo`. Besides, there should be no need to consider separately the case with no singularities or one singularity. One loop should suffice: `for s in singularities_list:`. It will automatically be skipped if the list is empty. That will also make the code cleaner.

Reply:
@jksuom  Sounds right and better. I'll implement it soon.
But I hope you understand that making the following cases is important (which I have done) :-
```
if singular_element == 0:

if singular_element == (len(singularities_list) - 1):

else
```
coz in the first case (`if singular_element == 0:`) the `lower_limit` has to be added
and the second case 
(`if singular_element == (len(singularities_list) - 1):`) the `upper_limit` has to be added
and the third case, 
rest of the singularities have to be added
and they can't be done in a single loop.

But, for sure, I'll implement this :
>no need to consider separately the case with no singularities or one singularity. One loop should suffice: `for s in singularities_list:`

Reply:
I don't think that the first or last item in the singularities list has any special significance. (Note that the list need not be ordered in any way.)